src/engine/memory.rs

```diff
@@ -0,0 +1,113 @@
+use std::ops::{Index, IndexMut};
+
+#[derive(Debug, Clone)]
+pub struct Memory<T> {
+    memory_size: usize,
+    segment_mask: usize,
+    segment_shift: u32,
+    default: T,
+    data: Vec<Vec<T>>,
+}
+
+impl<T: Copy> Index<usize> for Memory<T> {
+    type Output = T;
+    fn index(&self, index: usize) -> &Self::Output {
+        let segment_index = index >> self.segment_shift;
+        let segment_offset = index & self.segment_mask;
+        if self.data[segment_index].is_empty() {
+            return &self.default;
+        }
+        Index::index(&self.data[segment_index], segment_offset)
+    }
+}
+
+impl<T: Copy> IndexMut<usize> for Memory<T> {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        let segment_index = index >> self.segment_shift;
+        let segment_offset = index & self.segment_mask;
+        if self.data[segment_index].is_empty() {
+            self.data[segment_index] = vec![self.default; self.segment_mask + 1];
+        }
+        IndexMut::index_mut(&mut self.data[segment_index], segment_offset)
+    }
+}
+
+impl<T: Copy> Memory<T> {
+    pub fn new(default: T, memory_size: usize, segment_size: usize) -> Self {
+        assert!(
+            memory_size % segment_size == 0,
+            "memory size must be multiple of segment size"
+        );
+        assert!(
+            segment_size.is_power_of_two(),
+            "segment size must be a power of two"
+        );
+        let segment_mask = segment_size - 1;
+        let segment_shift = segment_size.trailing_zeros();
+        let segments = memory_size / segment_size;
+        Self {
+            memory_size,
+            segment_mask,
+            segment_shift,
+            default,
+            data: vec![[].to_vec(); segments],
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        self.data.iter().filter(|x| !x.is_empty()).flatten()
+    }
```

Comment:
That behaviour is kind of not, what I would expect. You are only iterating on allocated segments, no? I would expect to iterate over all elements in the address space.

Reply:
There are two parts in the answer to this:
1. Yes, I agree, the behavior is somewhat unexpected. But it is an essential part to the optimization because a majority of the time is wasted to iterating over uninitialized values just to filter them at the call site. I added some documentation to the function explaining the intended semantics. I hope this alleviates some of your concern.
2. Also note that the filtering of empty segments in this implementation does not change semantics. Even of all segments would be iterated over, the un-mapped segments would be empty and the flattened iterator would not cover any elements anyways. The early filtering is measurably faster though.

Reply:
Personally I would still prefer calling it `iter_mapped` or `iter_touched`, but I think at this point it comes down to taste. Anyway, with your added comment, I think it's fine!